### ISSUE TITLE
symbolic execution seems to go off track

### ISSUE THREAD

**Issue by bannsec** [angr/angr#196]
I have been working with rex and discovered a binary that rex doesn't successfully recognize the rip overwrite of a classic stack overflow. However, it also appears that angr doesn't successfully symbolically execute this binary, so that's probably the bigger problem and could be the reason why rex isn't working. Here's the code:

``` c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>

char buf[64];

void blerg() {
    char buf[8];
    read(0,buf,12);
}

int main() {
    char buf2[8];
    gets(buf2);
}
```

The make line is: `gcc -z execstack -fno-stack-protector -Wall -o stack_overflow stack_overflow.c`

The compiled binary I'm working with can be found here: http://pastebin.com/raw/HEysndYm

I'm following basic symbolic execution steps (import -> create state -> create path group -> explore) and it seems to go off track. Here's an example of the output trace:

``` python
In [7]: pg
Out[7]: <PathGroup with 7 active>

In [8]: pg.active[5]
Out[8]: <Path with 433 runs (at 0x106d1a3)>

In [9]: list(pg.active[5].trace)
Out[9]: 
['<SimIRSB 0x400470>',
 '<SimIRSB 0x400440>',
 '<SimProcedure __libc_start_main>',
 '<SimIRSB 0x4005b0>',
 '<SimIRSB 0x400400>',
 '<SimIRSB 0x400415>',
 '<SimIRSB 0x4005e1>',
 '<SimIRSB 0x4005e6>',
 '<SimIRSB 0x400540>',
 '<SimIRSB 0x40054b>',
 '<SimIRSB 0x4004e0>',
 '<SimIRSB 0x400518>',
 '<SimIRSB 0x4005fd>',
 '<SimIRSB 0x400606>',
 '<SimProcedure __libc_start_main>',
 '<SimIRSB 0x400587>',
 '<SimIRSB 0x400450>',
 '<SimIRSB 0x106ecc0>',
 '<SimIRSB 0x106ecda>',
 '<SimIRSB 0x106ecf4>',
 '<SimIRSB 0x106ed0a>',
 '<SimIRSB 0x106ed25>',
 '<SimIRSB 0x106ee20>',
 '<SimIRSB 0x107b2d0>',
 '<SimIRSB 0x107b2db>',
 '<SimIRSB 0x107b2ef>',
 '<SimIRSB 0x107b2fc>',
 '<SimIRSB 0x107b301>',
 '<SimIRSB 0x107b3e0>',
 '<SimIRSB 0x107b318>',
 '<SimIRSB 0x107b530>',
 '<SimIRSB 0x107a3d0>',
 '<SimIRSB 0x107a3da>',
 '<SimIRSB 0x107a3e8>',
 '<SimIRSB 0x107a5b0>',
 '<SimIRSB 0x107a5c4>',
 '<SimIRSB 0x107b490>',
 '<SimIRSB 0x107b4a0>',
 '<SimIRSB 0x107b4b7>',
 '<SimIRSB 0x106d130>',
 '<SimIRSB 0x106d145>',
....
```

It continues on like that. It appears to have jumped to a non-function part of the binary and is basically treating it as code, thus causing it to never finish and do strange things.

**Comment by zardus** [comment]
`0x0x1000000` should be where libc is loaded, so it sounds like it might be stuck in libc -- probably in gets. How long does it stay in libc?

**Comment by rhelmot** [comment]
Those addresses are in libc! We have no simprocedure for gets since there's no good way to emulate it symbolically, since the effect can cover literally the entire memory space. Or at least, that was the answer a year ago, since then we've implemented some options for the maximum amount of symbolic data a write can deal with normally. We should look into writing a good gets simprocedure.

So in some sense this is correctly emulating the binary, there is a literal infinite state explosion inside the libc binary.

**Comment by rhelmot** [comment]
(furthermore it's probably not emulating gets correctly at all since you're using an entry_state and not a full_init_state, which runs through the libc preinitializers, which set up some IO stuff)

**Comment by zardus** [comment]
In the rex usecase, when the standard input is constrained (and length-limited), it should be able to run through gets ok on its own, as long as the state is properly initialized :-)

**Comment by Owlz** [comment]
The rex use case is completely static afaik: `crash = rex.Crash("stack_overflow","AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA")`. Direct execution of that segfaults w/ rip overwrite. rex says that nothing happens and it's not a crash.

**Comment by rhelmot** [comment]
Ok, then this is technically a bug in rex, which should initialize its path group with `pg = proj.factory.path_group(proj.factory.full_init_state(...))`. This never really mattered for CGC since the full init state is the same as the entry state!

**Comment by Owlz** [comment]
Changing it to full_init_state didn't cause it to find the crash in this case (just tried it now).

**Comment by rhelmot** [comment]
Does it cause your testcase from the first post to terminate? You should also turn off LAZY_SOLVES for this

**Comment by Owlz** [comment]
For the angr side, it state splits like crazy with full_init_state. Likely won't stop (it was chewing up tons of memory).

**Comment by rhelmot** [comment]
Ok so this is just a general emulation problem in angr right now.

If you turn on the state option `STRICT_PAGE_ACCESS`, which causes a path to become errored if the emulator does something that should be a segfault, you can see that that our environment modeling simply isn't up to par, and somehow something isn't initialized correctly and it tries to write to a (concrete) address that isn't mapped.

We have a set of testcases that look for these issues in `angr/tests/test_accuracy.py`, which basically runs every arch through a hello world binary and makes sure that we never have any regressions in how far we are able to execute before things fall apart. So far, x86_64 is the only arch with a complete enough environment model to get through `puts("hello world");`, everything else needs syscalls implemented.

When we were getting started with the angr project, the big question that got tossed around a lot is "do we model the environment at the library level, or do we model it at the system level?" basically, do we implement simprocedures for every library function, or do we implement every syscall and make sure our loader is perfect?" I've been a big advocate for the latter, and test_accuracy.py is more or less the result of my initial work on that front. But really, the real answer to that big question is "neither, do your research on the CGC platform."

tl;dr something's wrong with angr's emulation. probably something with the brk syscall, or maybe I implemented a relocation incorrectly.

**Comment by ltfish** [comment]
> When we were getting started with the angr project, the big question that got tossed around a lot is "do we model the environment at the library level, or do we model it at the system level?" basically, do we implement simprocedures for every library function, or do we implement every syscall and make sure our loader is perfect?"

I was, and still am a big advocate for the former. Implementing `SimProcedures` for libraries can be done _on demand_, is usually easily debuggable, and are way more scalable than implementing syscalls (you can introduce different limits in individual library functions, or `SimProcedures`). That said, we do spend a lot of effort (mostly by @rhelmot) in correctly loading binaries and emulating syscalls.

Doing research on the CGC platform is a great plus - you don't have to worry about libraries unless you really want and have to. However, angr is also targeting both academic research and real-world binary analysis, so we do care about problems in binary loading, syscall/SimProcedure implementations, etc.

**Comment by frapik99** [comment]
Hi,
do you have an update on this? I'm facing a similar problem and I have just opened a new issue describing it.

Thanks

**Comment by techvoltage** [comment]
This slide(https://www.cs.umd.edu/~mwh/se-tutorial/symbolic-exec.pdf) suggests to use newlib, a trimmed down version of libc which can limit the complexity.
